velocity/src/main/java/com/proximyst/ban/event/subscriber/BannedPlayerJoinSubscriber.java

```diff
@@ -47,14 +47,14 @@ public void onJoinServer(final @NonNull LoginEvent event) {
 
     this.punishmentService.getActiveBan(event.getPlayer().getUniqueId())
         .join() // This *should* be fast, and only on one player's connection thread
-        .ifPresent(ban -> {
-          event.setResult(ComponentResult.denied(
-              this.messageService.formatMessage(ban.getPunishmentType()
-                      .getApplicationMessage(ban.getReason().isPresent()).orElse(null),
-                  ban)
-                  // We're about to deny them access; the time it takes to fetch the data doesn't matter.
-                  .join()
-          ));
-        });
+        .ifPresent(ban ->
+            event.setResult(ComponentResult.denied(
+                this.messageService.punishmentMessage(ban.getPunishmentType()
+                        .getApplicationMessage(ban.getReason().isPresent())
+                        .orElseThrow(() -> new IllegalStateException("application message must exist")),
+                    ban)
+                    // We're about to deny them access; the time it takes to fetch the data doesn't matter.
+                    .component().join()
+            )));
```

Comment:
The nesting makes this more difficult to understand than needed. I'd extract this into a separate method.